pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -121,28 +121,53 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 	// If the UserAccount has not been deleted, create or update user and identity resources.
 	// If the UserAccount has been deleted, delete secondary resources identity and user.
 	if !util.IsBeingDeleted(userAcc) {
+		reqLogger.Info("Adding finalizer on UserAccount")
 		// Add the finalizer if it is not present
 		if err := r.addFinalizer(userAcc); err != nil {
 			return reconcile.Result{}, err
 		}
-
+	}
+	if !util.IsBeingDeleted(userAcc) && !userAcc.Spec.Disabled {
+		reqLogger.Info("Ensuring user and identity associated with UserAccount")
 		var createdOrUpdated bool
 		var user *userv1.User
 		if user, createdOrUpdated, err = r.ensureUser(reqLogger, userAcc); err != nil || createdOrUpdated {
 			return reconcile.Result{}, err
 		}
-
 		if _, createdOrUpdated, err = r.ensureIdentity(reqLogger, userAcc, user); err != nil || createdOrUpdated {
 			return reconcile.Result{}, err
 		}
-
 		if _, createdOrUpdated, err = r.ensureNSTemplateSet(reqLogger, userAcc); err != nil || createdOrUpdated {
 			return reconcile.Result{}, err
 		}
-	} else if util.HasFinalizer(userAcc, userAccFinalizerName) {
-		if err = r.manageCleanUp(userAcc); err != nil {
-			return reconcile.Result{}, err
+	} else if util.HasFinalizer(userAcc, userAccFinalizerName) && util.IsBeingDeleted(userAcc) {
+		reqLogger.Info("Deleting user and identity associated with UserAccount")
+		deleted, err := r.deleteUserAndIdentity(userAcc)
+		if err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, userAcc, r.setStatusTerminating, err, "failed to delete user/identity")
+		}
+
+		// Remove finalizer from UserAccount
+		if deleted {
+			util.RemoveFinalizer(userAcc, userAccFinalizerName)
```

Comment:
So we don't remove the finalizer if there is nothing to delete?

We need:
1. If User OR Identity was deleted then just exit (the status is Terminating). Wait for the next reconcile.
2. If there is nothing to delete (no User or Identity exist). Then remove the finalizer. We are done.

Reply:
no, we remove the finalizer only after the user and identity have both been deleted. if we don't wait until after we will not enter into the same else-if condition. 

1. it does this, but the status is not set to "terminating" yet.. i will update this. 
2. yes it does this.

Reply:
what if there is nothing to delete? Identity & User do not exist. Then you won't delete the finalizer.

As I said we need to remove the finalizer when there is nothing to delete (so deleted==false). Why there is nothing to delete? Because it was just deleted in the previous reconcile loop or because it was never created we do not care. We can't remove the finalizer until we make sure user/identity are gone for good (vs. we just requested kube to delete them).

Reply:
no, if there is nothing to delete (everything has already been deleted) the finalizer will get removed. There is a test for this: https://github.com/codeready-toolchain/member-operator/pull/105/files#diff-fc74d9d211758dbe5bd817320a8fe2a1R904

Reply:
yeah looks good now